[Tee Ren Jing] iP

TeeRenJing opened the pull request:
## Ren

> "Kenough" Ken, Barbie 2023 [Source](https://www.independent.co.uk/arts-entertainment/films/news/ryan-gosling-ken-barbie-b2367042.htmlhttps://www.independent.co.uk/arts-entertainment/films/news/ryan-gosling-ken-barbie-b2367042.html)

- ~~Really~~ *nice*
- To **me**

1. Add your tasks
2. Find your tasks
3. Delete your tasks

- [x] Basic functions 👍 
- [ ] Add Colours
- [ ] Add ChatGPT


`main` method
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

TeeRenJing committed:
Rename, Greet, Exit

Changed Duke to Ren. Changed welcome message.

TeeRenJing committed:
Echo

Added in custom personality and user input and output functionality.

TeeRenJing committed:
Add, List

Added functionality to store tasks in a list. User may list out tasks with a command "list".

TeeRenJing committed:
Mark as Done

Added Functionality to mark and unmark tasks. Created new Task class and TaskList class as abstraction.

TeeRenJing committed:
Todos, Events, Deadlines

Added more types of Tasks (ToDo, Deadline, Event). Utilised Inheritance.

TeeRenJing committed:
Automated Text UI Testing

Configured runtest.bat file for automated UI testing on Windows.

TeeRenJing committed:
Handle Errors

Basic Error handling implemented. Added RenException to define exceptions specific to Ren.

TeeRenJing committed:
Delete

Added delete Task functionality.

TeeRenJing committed:
Use Enums

Used enums for Ren valid commands.

TeeRenJing committed:
Merge branch 'add-gradle-support' into branch-Level-7

TeeRenJing committed:
Add save to hard disk function

Saving the TaskList into local machine adds data persistence to Ren.

TeeRenJing committed:
Add retrieveFromHarddisk function

Adding this function completely ensures data persistence. State is saved in local hard disk and retrieved upon restarting Ren.

TeeRenJing committed:
Add retrieveFromHarddisk function

Adding this function completely ensures data persistence. State is saved in local hard disk and retrieved upon restarting Ren.

TeeRenJing committed:
Merge branch 'branch-Level-7'

TeeRenJing committed:
Merge branch 'branch-Level-7' of https://github.com/TeeRenJing/ip into branch-Level-7

TeeRenJing committed:
Add LocalDate support for Deadlines

Make Deadlines accept LocalDate

A LocalDate field is more appropriate compared to a String for the deadline and can be used for date and time related functionality.

TeeRenJing committed:
Merge branch 'branch-Level-8'

TeeRenJing committed:
Use more OOP

Add more java classes to stick to OOP encapsulation.

TeeRenJing committed:
Use More OOP

Added more classes for encapsulation, abiding to OOP.

TeeRenJing committed:
Organize into packages

Preliminary attempt to organize code into packages to refactor codebase.

TeeRenJing committed:
Use Gradle

Use Gradle as a tool for build automation.

TeeRenJing committed:
Add JUnit Tests

Added simple JUnit tests for ToDo and Deadline

TeeRenJing committed:
Create a JAR file

Adds and refactor code to enable packaging app as an executable JAR file.

TeeRenJing committed:
Add Find functionality

Allows user to find for certain tasks.

TeeRenJing committed:
Merge branch 'branch-Level-9'

TeeRenJing committed:
Add JavaDoc comments

TeeRenJing committed:
Merge branch 'branch-A-JavaDoc'

TeeRenJing committed:
Follow the Coding Standard

TeeRenJing committed:
Merge branch 'branch-A-CodingStandard'

TeeRenJing committed:
Merge branch 'branch-Level-7'

TeeRenJing committed:
Merge branch-Level-8

TeeRenJing committed:
Use CheckStyle

TeeRenJing committed:
Merge branch 'branch-A-CheckStyle'

tanteckfang commented on src/main/java/Ren/InsufficientArguments.java:
```diff
@@ -0,0 +1,7 @@
+package ren;
+
+public class InsufficientArguments extends RenException {
```
Alternative class name: InsufficientArgumentsException

tanteckfang commented on src/main/java/Ren/RenObjectMapper.java:
```diff
@@ -0,0 +1,112 @@
+     *
+     * @param taskList TaskList to be stored
+     */
+    public static void storeIntoHarddisk(TaskList taskList) {
```
👍 verb is used for naming method

tanteckfang commented on src/main/java/Ren/task/Deadline.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+
+    public Deadline(String deadlineDesc, boolean isDone, TaskType taskType, java.time.LocalDate deadline) {
```
able to understand from variable name

tanteckfang commented on src/main/java/Ren/task/TaskList.java:
```diff
@@ -0,0 +1,143 @@
+     * @return the toggled task
+     */
+    public Task toggleTask(String[] commandArr) {
+        Task task = this.tasks.get(Integer.parseInt(commandArr[1]) - 1);
```
👍 able to distinguish clearly between single-valued and multi-valued variables

TeeRenJing committed:
Merge branch 'branch-Level-8'

TeeRenJing committed:
Fix list formatting in GUI

TeeRenJing committed:
Merge branch 'branch-Level-10'

kenvynKwek commented on src/main/java/Ren/Commands.java:
```diff
@@ -0,0 +1,28 @@
+ * Represents the commands that the user can input
+ */
+public enum Commands {
+    LS_COMMAND("list"),
+
+    FIND_COMMAND("find"),
+
+    DELETE_COMMAND("delete"),
+
+    EXIT_COMMAND("bye"),
+
+    MARK_COMMAND("mark"),
+
+    UNMARK_COMMAND("unmark");
```
Should there be blank lines in between each *_COMMAND enum constant/value ?

kenvynKwek commented on src/main/java/Ren/Ren.java:
```diff
@@ -0,0 +1,44 @@
+
+import java.util.ArrayList;
+import java.util.Scanner;
+
```
Do consider having a header comment to briefly explain what the Ren class does

kenvynKwek commented on src/main/java/Ren/RenParser.java:
```diff
@@ -0,0 +1,58 @@
+            return "Bye. Hope to see you again soon!";
+        }
+        String[] commandArr = inputStr.split(" ");
+        if (commandArr[0].equals(Commands.LS_COMMAND.getValue())) {
+            return tasks.toString();
+        } else if (commandArr[0].equals(Commands.FIND_COMMAND.getValue())) {
+            return tasks.filterToListOfMatchingTasks(commandArr[1]).toString();
+        } else if (commandArr[0].equals(Commands.DELETE_COMMAND.getValue())) {
+            try {
+                Task task = tasks.deleteTask(commandArr);
+                return task + " deleted";
+            } catch (InsufficientArgumentsException e) {
+                RenUi.printException(e);
+            }
+        } else if (commandArr[0].equals(Commands.MARK_COMMAND.getValue())
+                || commandArr[0].equals(Commands.UNMARK_COMMAND.getValue())) {
+            Task task = tasks.toggleTask(commandArr);
+            if (commandArr[0].equals(Commands.MARK_COMMAND.getValue())) {
+                return task + " marked as done";
+            } else {
+                return task + " marked as undone";
+            }
+        } else if (TASK_TYPES.contains(commandArr[0])) {
+            try {
+                Task task = tasks.addTask(commandArr);
+                return task + " added";
+            } catch (InsufficientArgumentsException e) {
+                RenUi.printException(e);
+            }
+        } else {
+            return "Please enter a proper command";
+        }
```
Could a switch-case statement be used here instead ? (maximize readability)

kenvynKwek commented on src/main/java/Ren/task/Deadline.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+
+    public Deadline(String deadlineDesc, boolean isDone, TaskType taskType, java.time.LocalDate deadline) {
+        super(deadlineDesc, isDone);
+        this.DEADLINE = deadline;
+        this.taskType = TaskType.DEADLINE;
+    }
+
+    public Deadline(String deadlineDesc, boolean isDone, java.time.LocalDate deadline) {
+        super(deadlineDesc, isDone);
+        this.DEADLINE = deadline;
+        this.taskType = TaskType.DEADLINE;
+    }
+
+    public Deadline() {
+        super();
+        this.DEADLINE = java.time.LocalDate.now();
+        this.taskType = TaskType.DEADLINE;
+    }
```
Excellent use of Method Overloading for Deadline constructors !

kenvynKwek commented on src/main/java/Ren/task/Event.java:
```diff
@@ -0,0 +1,61 @@
+    private final String START_TIME;
+    private final String END_TIME;
+
+    public String getDescription() {
+        return this.description;
+    }
+
+    public boolean getIsDone() {
+        return this.isDone;
+    }
+
+    public TaskType getTaskType() {
+        return this.taskType;
+    }
+
+    public String getSTART_TIME() {
+        return this.START_TIME;
+    }
+
+    public String getEND_TIME() {
+        return this.END_TIME;
+    }
```
Could the tell-don't-ask OOP principle be employed here instead of getter methods ?

kenvynKwek commented on src/main/java/Ren/task/TaskList.java:
```diff
@@ -0,0 +1,157 @@
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.List;
+
```
Do consider a header comment to briefly explain what the TaskList class does

kenvynKwek commented on src/main/java/Ren/task/Task.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    protected String description;
+    protected boolean isDone;
```
boolean named to sound like booleans. Nice !

kenvynKwek reviewed (commented):
Overall, your code is easy to read and understand! I noticed you used getter methods for classes under the task package, perhaps usage of the tell-don't-ask OOP principle could be applied and should ease integrating future implementations with Task objects.

TeeRenJing committed:
Use Assertions

Use assertions in codebase.

Assertions help us ensure various important assumptions and invariants hold at certain points in the code.

Let's,
* add assertions to assert file is present and length of TaskList is non negative.

TeeRenJing committed:
Improve Code Quality

Improve code quality by refactoring.

There are some methods where the code is too nested. This reduces readability of the codebase.

Reducing the nesting in functions improves the code quality.

Let's,
* extract long nested code into methods for better readability.

TeeRenJing committed:
Merge pull request #1 from TeeRenJing/branch-A-Assertions

Use Assertions

TeeRenJing committed:
Merge branch 'master' into branch-A-CodeQuality

TeeRenJing committed:
Merge pull request #2 from TeeRenJing/branch-A-CodeQuality

Refactor codebase for better code quality

TeeRenJing committed:
Implement case insensitive search

TeeRenJing committed:
Merge branch 'branch-BCD-Extension'

TeeRenJing committed:
Set up a product website

TeeRenJing committed:
Merge branch 'branch-A-UserGuide'

TeeRenJing committed:
Fix issues with Jar file

TeeRenJing committed:
Merge branch 'branch-A-Release'

TeeRenJing committed:
Update README.md

TeeRenJing committed:
Update README.md

TeeRenJing committed:
Refactor and add unit tests

TeeRenJing committed:
Add title and proper exit behavior

TeeRenJing committed:
Update exit behavior
